리사이클러뷰 어댑터 코드개선

YunSeokVV opened the pull request:
### PR 간단하게 설명 해주세요.
- #7 에서 언급해주신 리사이클러뷰 관련 수정사항들 참고후 수정하였습니다.
- Retrofit 관련해서 첨언해주신 것은 아직 적용하지 못했습니다.
- SourceTree 를 다루다가 실수를 너무 많이 했네요. 로컬에 있던 멀쩡한 프로젝트를 지우고 복구 하는 작업을 너무 많이 했습니다. 앞으로 이런 실수가 없도록 하겠습니다.

YunSeokVV committed:
test

YunSeokVV committed:
Revert "test"

This reverts commit bc163be9e9b62aa1f3be6c31bf78967031d59bbf.

f-lab-loco commented:
수정하신게 없네요 Files changed에서 확인 부탁드립니다.

YunSeokVV commented:
PR전부 수정한뒤 새로운 Repository생성후 Retrofit프로젝트를 푸시후 링크 드리도록 하겠습니다.

YunSeokVV committed:
체크용브랜치 체크

YunSeokVV committed:
- "https://github.com/YunSeokVV/AndoridPracticeKotlin/pull/9" (리사이클러뷰 어댑터 코드개선 #9) 에서 누락됐던 리사이클러뷰 어댑터 관련 수정사항들 적용
- retrofit, Glide 라이브러리 사용해서 리사이클러뷰 이미지 표현
- println 로그 전부 삭제

YunSeokVV committed:
Merge branch 'feature/recyclerView' of https://github.com/YunSeokVV/AndoridPracticeKotlin into feature/recyclerView

YunSeokVV committed:
테스트용 커밋

YunSeokVV committed:
commit

YunSeokVV committed:
test

YunSeokVV committed:
d

YunSeokVV committed:
Retrofit 헤더 어노테이션을 사용해서 적용

YunSeokVV committed:
- kapt 적용
- 무의미하게 lateinit var 로 선언된 변수들 let 또는 let by lazy로 선언
- RoomDB 에 관해서 공부
-타이틀바의 필터 이미지 오류 수정
-사용자가 저장한 식물 정보가 있는 경우 my garden화면 수정
-Logger 라이브러리 추가

YunSeokVV committed:
(변경 파일을 staged 하지 않았기 때문에 다시 푸시함)
- kapt 적용
- 무의미하게 lateinit var 로 선언된 변수들 let 또는 let by lazy로 선언
- RoomDB 에 관해서 공부
-타이틀바의 필터 이미지 오류 수정
-사용자가 저장한 식물 정보가 있는 경우 my garden화면 수정
-Logger 라이브러리 추가

f-lab-loco commented on app/build.gradle.kts:
```diff
@@ -9,12 +10,15 @@ android {
     defaultConfig {
         applicationId = "com.example.samplekotlin"
-        minSdk = 24
+        minSdk = 26
```
왜 minSdk 를 올리신건가요?

YunSeokVV replied:
kapt 를 적용하는 과정에서 컴파일 에러가 발생해서 올렸습니다. 정확히 어떤 어떤 에러가 발생했을때 minSdk 를 올려야 한다는 내용을 봤었는지 기억하지 못했습니다. 다음번 부터는 버전관련해서 수정사항이 생기면 어떤 이유 때문에 수정을 하게 됐는지 따로 메모를 해놓도록 하겠습니다.
다시 minSdk 를 24로 수정해서 진행하는데 앱이 빌드되는데는 아무런 이상이 없습니다.

안드로이드에서 말하는 API가 무엇인지 공부를 했습니다.
감사합니다.

f-lab-loco commented on app/build.gradle.kts:
```diff
@@ -27,21 +31,33 @@ android {
     androidTestImplementation("androidx.test.espresso:espresso-core:3.5.1")
+
+    // 로그를 보기 편하게 해주는 라이브러리
+    implementation("com.orhanobut:logger:2.2.0")
```
Timber 도 찾아보셨나요? logger 가 더 편하셨나요?

YunSeokVV replied:
Timber에 관해서 리서치를 진행했습니다. Timber가 logger 비해서 조금 더 커스텀 할 수 있는 것들이 많습니다.
(Debug나 Release 모드일때만 로그를 찍게 하기), Timber 클래스를 상속 받아서 커스텀하기
그런데 저는 logger 개인적으로 가독성이 좋다고 느껴져서 일단 이번 프로젝트에서는 logger를 사용해보도록 하겠습니다.
좋은 피드백 해주셔서 감사합니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/adpater/PlantListAdapter.kt:
```diff
@@ -6,34 +6,53 @@ import android.view.ViewGroup
 
 class PlantListAdapter (private val data: List<Plant>) : RecyclerView.Adapter<PlantListAdapter.ViewHolder>(){
 
+    private  lateinit var listener : OnItemClickListener
```
여전히 lateinit 을 쓰셨는데 제가 보기엔 장점보단 단점이 많은 키워드 아닌가요?
장점과 단점 설명해줄 수 있어요?

YunSeokVV replied:
장점 : 
1. 나중에 객체를 바꿔서 넣을 수 있다.
2. 객체가 Null이 될 수 없다. (객체를 사용하기전에 초기화 해야한다. 안그러면 런타임에서 에러 발생)

단점 : 
1. 원시타입을 넣을 수 없다.

ps. 제가 놓친 부분인 것 같습니다. 이번 커밋내역에서 val 로 수정하도록 하겠습니다. 감사합니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/adpater/PlantListAdapter.kt:
```diff
@@ -6,34 +6,53 @@ import android.view.ViewGroup
+            plantName.text = item.name
+            Glide.with(itemView.context).load(item.imageResource).into(plantImage)
+            // 이미지뷰의 모서리를 약간 둥글게 만든다.
+            plantImage.clipToOutline = true
```
round 때문에 넣으신거라면 Glide 에는 round 처리하는 옵션 없을까요?

YunSeokVV replied:
Glide에서 round 처리를 하는 옵션을 몰랐는데 이번 기회에 알게 됐습니다.
원본 SunFlower프로젝트를 보면 좌측하단과 우측상단의 모서리만 약간 둥글게 처리가 되어 있습니다.
모서리중에서 특정 부분만 둥글게 만드는 방법은 Glide에서 할 수 없다고 판단했습니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/database/PlantDao.kt:
```diff
@@ -0,0 +1,15 @@
+@Dao
+interface PlantDao {
+    @Query("SELECT * FROM Plant")
+    fun getAll():List<Plant>
```
이렇게쓰면 동기로 select 받나요? 비동기로 받나요?
그리고 동기와 비동기에 대해서 차이 설명해주세요.

YunSeokVV replied:
비동기적으로 select를 받습니다. 서브 스레드에서는 DB와 관련된 작업을 진행하는것과 무관하게 main스레드는 ui관련 작업을 계속 진행하고 있는 상황이고 앱을 사용하는 유저는 다른 기능들을 활용할 수 있습니다.

동기라는 것은 작업이 끝난 뒤에 다른 작업을 진행하는 방법을 의미합니다.
예로 군대에서 사격 훈련을 할때 사격조는 부사수로부터 반드시 탄알집을 인계받아야지만 사격이 가능합니다. 탄알집을 인계받기 전에는 사격을 할 수 없습니다. 
그와 반대로 비동기는 특정 작업이 끝나는 것과 무관하게 여러 작업이 진행되는 것을 의미합니다.
예로 소대장은 병사들의 탄알집을 불출하기 위해서 탄창에 총알을 채워 넣는 작업을 하는데 이 작업은 사격조나 부사수가 하는 작업과 무관하게 계속해서 이루어 집니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/database/PlantDatabase.kt:
```diff
@@ -0,0 +1,37 @@
+        private var instance: PlantDatabase? = null
+
+            fun getInstance(context : Context): PlantDatabase? {
+                if(instance == null ){
```
이건 singleton 중 무슨패턴이죠?
singleton 만드는 방법 여러개가 있는데 설명해주세요.

YunSeokVV replied:
위의 방법은 companion Obeject로 싱글톤 패턴을 구현하는 방법중 하나입니다.
코틀린에서 싱글톤 패턴을 구현하는 방법은 두가지 방법이 있습니다.
companion object또는 object 키워드를 사용하는 방법입니다.
하나씩 설명을 해보겠습니다.
companion object 는 주로 비동기 작업을 수행할 때 객체가 멀티스레드로부터 접근 되는 상황을 방지할 때 사용하는 싱글톤 객체 생성 방법입니다. 클래스 안에서 정의합니다.

private var instance: PlantDatabase? = null

저는 위와 같은 방법으로 객체를 선언했었는데요. 제가 여기서 companion object를 활용한 방법 대로라면 멀티스레드로부터 객체의 접근을 보호할 수 없기때문에 굳이 객체를 var로 선언해야 할 필요가 없습니다.
@Volatile 어노테이션과 synchronized 예약어를 사용했으면 변수를 위와 같이 선언 하는게 맞습니다.

두번째 방법은 object를 사용하는 방법입니다.object를 사용해서 싱글톤 객체를 만드는 방법은 비동기 작업에서 멀티스레드로부터 객체를 보호해주지 못합니다. 주로 익명 객체를 만들때 사용되는데요. 이는 객체 자체를 해당 클래스에서만 사용하고 다른 곳에서는 다시 쓸 일이 없을 때 사용하는 방법입니다. 

RoomDB 를 리서치했을때 제 앱이 멀티스레드 환경으로 구성된게 아니라면 싱글톤 객체로 PlantDatabase 를 구성하라고 되어 있었습니다. 감사합니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/view/MainActivity.kt:
```diff
@@ -11,14 +11,14 @@ import androidx.viewpager2.widget.ViewPager2
 
 class MainActivity : AppCompatActivity() {
-
+    var filterVisible: Boolean = false
```
접근제어자 신경 써주세요. 기본적으로 private 을 붙이는 습관 들여주세요.

YunSeokVV replied:
네 알겠습니다.
앞으로 반드시 습관을 들이도록 하겠습니다. 여태까지 멘토님께서 습관을 들이라고 말씀해 주셨던 것들을 한번 상기해보겠습니다.

1. 멤버변수 앞에 m 을 붙이지 않기. 옛날 자바 개발 방식임.
2. 모든 자바기반 개발을 할때는 cammelCase로 이름을 지정하기.
(공부를 해본 결과 json 값은 통상적으로 snakeCase 를 사용하는 것을 알게됐습니다. 반드시 준수해야 한다는 것은 아니지만.. 자바처럼 약속이죠)
3. 특별한 이유가 없다면 반드시 private 접근제어자를 사용할것.
4. 코드를 반드시 주기적으로 정렬할것.

제가 멘토님께 피드백 받은 습관들은 위와 같습니다. 혹시 제가 놓치고 있는 부분이 있다면 첨언 부탁드리겠습니다.
훌륭한 습관을 들이도록 도와주셔서 감사합니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/view/MyGardenFragement.kt:
```diff
@@ -1,26 +1,61 @@
+import java.util.concurrent.Executors
 
 class MyGardenFragement : Fragment() {
+    val executor: Executor = Executors.newSingleThreadExecutor()
```
Executor 종류는 어떤게 있는지 알고 계신가요?
executor 와 thread 의 차이도 알고 계신가요?

YunSeokVV replied:
자바에서 executor 의 종류는 아래 5가지 정도가 있는 것으로 확인했습니다.
1.ThreadPoolExecutor 
2.ScheduledThradPoolExecutor 
3.SingleThreadExecutor 
4.CachedThreadPool 
5.FixedThreadPool

executor 에서는 ThreadPoolExecutor 를 사용해서 ThreadPool안의 스레드를 관리해주지만 Thread 단독으로 해당 기능을 사용할수 없습니다.
thread에서는 스케쥴 작업에 대한 기능을 제공해주지 않지만 executor에서는 해당 기능을 제공해줍니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/database/PlantDatabase.kt:
```diff
@@ -0,0 +1,37 @@
+    companion object {
+        private var instance: PlantDatabase? = null
+
+            fun getInstance(context : Context): PlantDatabase? {
```
싱글톤인데 nullable 하게 객체가 반환되네요?
그리고 nonnull하게 !! 로 하지마시고 코틀린키워드를 사용해서 nonunull로 반환되게 수정해주세요.

YunSeokVV replied:
멘토님 우선 제가 private var instance: PlantDatabase? = null 로 코딩한 부분은 수정하도록 하겠습니다. 말씀해주신 대로 싱글톤 패턴을 nullable 하게 구현한 부분에서 우선 제가 첫번째 잘못을 했네요. 
아래 코드는 멘토님이 말씀해주신 nonnull로 반환하는 코드입니다.
```
private var instance: PlantDatabase? = null
    fun getInstance(context : Context): PlantDatabase? {
        instance ?:synchronized(PlantDatabase::class){
            instance = Room.databaseBuilder(
                context.applicationContext, PlantDatabase::class.java, "plant_database").build()
        }
        return instance
    }
```

그리고 아래의 코드는 지금와서 생각해보니 애초에 아래와 같은 방식으로 짜는게 맞았다는 생각이 들어서 남기는 코드입니다. 
```
companion object {
        fun getInstance(context : Context): PlantDatabase {
                //다른 메소드가 접근하는 것을 허용하지 않게끔 lock 시켜주고 하나의 스레드만 접근할 수 있게끔 해준다.
                synchronized(PlantDatabase::class){
                    val instance = Room.databaseBuilder(
                        context.applicationContext, PlantDatabase::class.java, "plant_database").build()
                    return instance
                }
        }
}
```

ps. 싱글톤 패턴은 객체 하나만을 공유해서 사용하는 방식입니다. 다른 클래스에서 서로다른 객체를 만들어서 쓰는 방법이 아닙니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/view/PlantListFragment.kt:
```diff
@@ -1,69 +1,131 @@
+    val plantNames = arrayOf("Apple", "Avocado", "Beet", "Bougainvillea", "Cilantro", "EggPlant", "Grape", "Hibiscus", "Mango", "Orange", "Pear", "Pink & White", "Rocky Mountain", "Sunflower", "Tomato", "Watermelon", "Yulan Magnolia")
+    //val plantImages = arrayOf("apple", "avocado", "beet", "bougainvillea", "cilantro", "eggplant", "grape", "hibiscus", "mango", "orange", "pear", "pink_white", "rockymountian", "sunflower", "tomato", "watermelon", "yulanmagnolia")
+    // 각 꽃의 고유 아이디들(unSplash 사이트에서 확인가능! 마지막 목련은 그냥 id값으로 줬다. 구할 수 없었기 때문)
+    val plantImagesID = arrayOf("oo3kSFZ7uHk", "cueV_oTVsic", "TmjyLCUpcDY", "CdkGk6l4eCM", "FVikKhvhnKE", "8cqlBGw84oU", "vHnQqO3TT4c", "yFmwdScAUS8", "2vq33LK8bZA", "A4BBdJQu2co", "cfxnOUSLrgk", "fxbzT96KzOo", "7CHMu7uPHCQ", "2IzoIHBgYAo", "5Oi5sG6G0z8", "hUkZv0Y47Ic", "pDGNBK9A0sk")
```
검색 api를 사용해서 plant의 id를 알순 없나요??

YunSeokVV replied:
공부를 해본결과 unsplash api에서는 이미지에 대한 고유 ID 값은 공식문서에서 따로 확인할 수 있는 방법을 제공해주지 않는것으로 확인했습니다. 따라서 plant의 id 값도 공식문서를 참고해서는 알 수 없다는 결론에 이르렀습니다.
GPT에 검색을 해서 이미지의 고유 ID 값을 확인할 수 있는 요청값을 알아냈습니다.

이렇게 공식문서에 없는 스펙을 undocumented API 라고 하는 사실을 알아냈습니다.
이걸 알아내려면 개발자가 여러가지 실험을 해봐야 알 수 있다고 하네요...
감사합니다.

YunSeokVV replied:
정정하겠습니다. 공식도큐먼트에 이미의 고유 id 값 없이 그냥 이미지명으로 검색하는 방법이 있었는데 제가 확인을 제대로 안했었네요. 해당 방법을 사용해서 구현해보도록 하겠습니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/vo/URLs.kt:
```diff
@@ -0,0 +1,10 @@
+import com.google.gson.internal.LinkedTreeMap
+
+data class URLs(
+    @field:SerializedName("urls")
```
SerializedName 안 붙으면 어떤일이 발생하나요?

YunSeokVV replied:
어노테이션을 붙이지 않아도 잘 동작하는 것으로 확인하였습니다.
왜 이게 가능한 것인지 찾아봤습니다.
data class의 경우 getter와 setter를 자동으로 생성해주는데(변수가 val 일때는 getter만 var 일때는 setter까지 추가해줍니다.) 자동으로 생성된 getter메소드에 SerializedName 어노테이션이 붙기에 가능했던 것입니다.

물론 어노테이션을 안붙여도 기능이 동작하는데 상관은 없지만 가독성과 코드의 명확한 이해를 위해서 어노테이션을 붙이는게 좋다는 정보를 찾았습니다.
감사합니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/vo/URLs.kt:
```diff
@@ -0,0 +1,10 @@
+
+data class URLs(
+    @field:SerializedName("urls")
+    val urls: LinkedTreeMap<String, Any>
```
LinkedTreeMap 이 어떤 컬렉션인지 설명가능해요?
같이 비교군으로 나오는 Map이 어떤게 있나요?

YunSeokVV replied:
LinkedTreeMap 은 자바 컬렉션 프레임워크에 존재하지 않는 자료구조입니다. (리서치를 하다가 이 부분에서 좀 놀랐네요)
LinkedTreeMap 은 구글에서 제공해주는 GSON 라이브러리 안에 존재하는 자료구조입니다.
JSON값을 파싱하는데 특화되어 있습니다.
JSON값의 key-value 쌍이 순서대로 정렬되어 있을때 linkedList 의 장점(순서대로 값이 정렬된 상황에서 데이터를 탐색하는데 효과적입니다.)을 활용해서 값을 파싱하는데 좋습니다.

그리고 LinkedTreeMap 을 공부하기 위해서 살펴본 자료구조들에 대해서 말씀드리겠습니다.
Tree : 비선형구조(저장된 데이터의 앞뒤 순서가 무관한 경우)로 데이터를 저장하며 부모 노드와 자식 노드가 존재합니다.
파일 디렉터리나 앞서 말한 JSON 값이나 XML을 담는데 용이합니다.
이진트리에서 저장되는 데이터가 편협하게 한쪽에만 몰리게 되면 데이터를 탐색하는 속도가 느려질 수 있는데 이를 보완하기 위해서 Red Black Tree 와 같은 자료구조가 만들어 진 것으로 공부하였습니다.

Linked List : 연결된 리스트라고도 부르는데 리스트와는 다르게 값을 저장할때 인덱스를 사용하지 않습니다. 그냥 값을 저장할때 데이터 간의 전후 관계에만 의존한다는 특징이 있습니다.
예로 데이터를 저장한다고 가정하면
"라면 -> 짜장면 -> 치킨 -> 피자" 
라는 순서로 저장을 하게 되는데 만약 제가 이 linked list 안에서 "치킨" 을 찾고 싶다면 앞의 내용들을 알아야 뒷내용을 알기 때문에 처음부터 끝까지 찾는 것 이외에는 방법이 없습니다.

f-lab-loco reviewed (commented):
코드 자동 정렬 하면서 코드 짜주세요. 이 요청도 저번에 말씀드린 것 같습니다.

YunSeokVV commented:
> 코드 자동 정렬 하면서 코드 짜주세요. 이 요청도 저번에 말씀드린 것 같습니다.

네 알겠습니다. 앞으로 습관을 들이도록 할 것이고 다음 수업시간에는 정렬된 코드를 보실 수 있도록 준비하겠습니다.
좋은 지적 감사합니다.

YunSeokVV committed:
-github에서 받은 피드백들 적용
-+버튼을 누른뒤 보이지 않게끔 처리

YunSeokVV committed:
3주차수업때 받은 피드백들 적용
+버튼 누른뒤 보이지 않게끔 처리

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/adpater/PlantListAdapter.kt:
```diff
@@ -10,24 +10,27 @@ import com.bumptech.glide.Glide
+class PlantListAdapter(private val data: MutableList<Plant>) :
+    RecyclerView.Adapter<PlantListAdapter.ViewHolder>() {
+    //todo: by lazy를 사용해서 늦은 초기화를 하고 싶었는데 어떻게 해야할지 도저히 떠오르지가 않는다.
+    private lateinit var listener: OnItemClickListener
```
결론은 생성자에서 추가 안하시는게 답이나오셨나요?
그럼 var로 선언하는것 외엔 없을 것 같습니다.

YunSeokVV replied:
관련해서 질문드립니다.
1. 클릭이벤트를 인터페이스를 val 로 선언하기 위해서 Adapter클래스의 생성자에 context까지 넣어가면서 구현해야 할 가치가 있는가? 
(context는 viewHolder의 itemView에서 제공해준다고 제게 가르쳐주셨습니다. 그런데 익명 객체안에서 conetxt를 사용하기 위해서는 생성자나 set을 이용해서 context를 설정하고 사용하는 것 이외에는 방법이 없겠더군요.)
2. 클릭이 일어났을때 수행해야 할 동작이 같은 경우에는 사진과 같은 방법으로 해도 문제가 될 것은 없지 않을까? 
(리사이클러뷰의 아이템 클릭이 발생했을 때 수행해야할 동작이 좋아요를 누른 식물목록을 표현하는 Fragment나 전체 식물목록을 표현하는 Fragment나 같다면 사진과 같은 방법으로 구현하는 것도 나쁘지 않은 방법이라고 생각합니다. 기존의 방식대로라면 두개의 Fragment에 각각 listener 익명객체를 구현했었기 떄문입니다. 중복되는 소스라면 두개보다는 하나가 더 낫지 않나 싶네요.)

어댑터에 아래와 같이 인터페이스 객체를 초기화한뒤 생성자에서 context를 받아서 동작시키면 클릭이벤트가 동작은 잘 됩니다.

```
    public val listener2 = object : OnItemClickListener {
        override fun onItemClick(plant: Plant) {
            Logger.v("onItemClick called")
            val intent: Intent = Intent(context, PlantDetailActivity::class.java)
            intent.putExtra("plantObject", plant)

            context.startActivity(intent)
        }
    }
```

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/util/ExecutorInterface.kt:
```diff
@@ -1,4 +1,18 @@
 interface ExecutorInterface {
+
+    fun executeApi(doThing: Any) {
+        val executor: Executor = Executors.newSingleThreadExecutor()
```
interface 에 구현하시는게 일반적인 추상화 이론인가요?
해당 인터페이스의 상속은 시도해보셨나요?

YunSeokVV replied:
interface를 구현하는 것은 일반적인 추상화 이론인 것으로 공부하였습니다.
자바에서 말하는 추상화는 중복되는 내용들을 최대한 한대 모아서 관리하는 것이라고 할 수 있습니다.
기본적으로 부모 클래스에서 중복되는 메소드나 변수들을 모아서 관리하는 것도 추상화라고 할 수 있고 추상 클래스와 인터페이스를 사용하는 것도 추상화를 적용했다고 얘기할 수 있습니다.

이 인터페이스를 상속하는 시도는 따로 해보지 않았습니다.
대신 인터페이스의 구현체를 만들어서 뷰단에서 사용하려고 했었지만 의도대로 동작은 되지 않았습니다.

그리고 첨언해서 인터페이스의 추상메소드에 대한 명확한 이해가 없는 상대로 사용하였습니다.
인터페이스의 추상메소드는 메소드의 내용을 비워놓아야 합니다.
저는 일정양 내용을 채워놨는데 이 경우는 default 메소드라고 합니다.
deafult 메소드는 자바8(1.8)부터 추가 되었고 인터페이스의 추상메소드가 늘어나게 됐을때 구현한 객체가 추상 메소드를 전부 구현해줘야 하는 단점을 보완하기 위해서 만들어진 기술입니다.
(사실 solid 의 인터페이스 분리의 원칙을 잘 지켰다면 애초에 이런 일이 일어나지도 않을텐데.. 라는 생각이 드네요.)

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/util/Util.kt:
```diff
@@ -1,4 +1,11 @@
+import android.widget.Toast
+
+class Util(context: Context) {
+    private val context: Context = context
```
이렇게 쓰신건 왜 쓰셨나요? 기존에는 생성자에 `class Util(private val context: Context)` 로 쓰지 않으셨나요? adapter code

YunSeokVV replied:
말씀해주신 방법이 더 나은 방법인게 맞습니다. 자바에서 쓰던 내용을 코틀린에서 축약하는 여러가지 방법들을 제공해주는데 아직 제가 익숙하게 사용하는 단계는 아닌듯 합니다. 지적해주셔서 감사합니다.

f-lab-loco reviewed (commented):
예제코드 지워지면 그때 다음 depth로 나가시죠.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/util/ExecutorInterface.kt:
```diff
@@ -1,4 +1,18 @@
+//뷰 단에서 이 인터페이스를 실행시키고 매개변수만 넘겨줘서 할 일을 넘겨주면 DB에 데이터를 저장하거나 레트로핏을 사용해서 통신을 하는게 가능할 줄 알았는데 안됐습니다.
 interface ExecutorInterface {
+
+    fun executeApi(doThing: Any) {
```
람다함수 공부하셨으니까 적용

YunSeokVV replied:
람다함수를 적용은 했습니다.
인터페이스를 활용해서 기능을 구현하려고 했는데 공부를 하고나서 두가지 방법이 있는 것을 알게됐습니다.
1.인터페이스의 디폴트 메소드를 익명 객체로 만들어서 인터페이스 메소드 기능을 사용하기
2.인터페이스를 구현하는 클래스를 만들고 그 클래스를 기반으로 객체를 생성해서 추상화하기

두가지 다 해봤는데 원래 목적인 코드를 축약하는 것에는 실패를 했네요. 그래도 싱글톤을 적용해서 executor 객체가 한번만 선언되도 된다는 점과 람다에 관해서 조금 더 친숙해졌다는데 의의를 두도록 하겠습니다.
감사합니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/util/Util.kt:
```diff
@@ -1,4 +1,11 @@
+import android.content.Context
+import android.widget.Toast
+
+class Util(context: Context) {
```
1. object 로 선언해서 Util -> 싱글톤
2. companion object 로 선언해서 Util -> static function
3. 윤석님 코드

즉, class Util1, class Util2, class Util3 개발해서 차이 설명 부탁드릴게요.
선언 및 실행까지 해보는걸로 부탁드려요.

YunSeokVV replied:
세가지 방법을 다 써봤습니니다. 일단 제가 혼자 썼던 기존의 방법은 객체의 메소드를 사용하려고 할때마다 객체를 초기화 해서 같은 코드가 중복되는 문제가 있었습니다. 더 나쁜 방법이였던거죠.

Util1의 경우 (object를 사용해서 싱글톤 구현)
object를 사용해서 인스턴스를 만들어서 구현하는 경우에는 선언과 동시에 객체가 메모리에 올라가게 됩니다.
그리고 object로 싱글톤을 구현하면 멀티스레드 상황일때 대처가 불가능한데 저는 구현 상황이 멀티스레드가 아니기 때문에 이 부분은 지금 상황에 중요하지 않을 것 같습니다.

Util2의 경우 (companionObject를 사용해서 싱글톤 구현)
클래스 내부에서 제가 구현하고 싶은 기능들을 구현했습니다. 
처음 공부를 했을때 객체의 instance를 lateinint var 로 선언하고 사용되기전에 해당 값이 null인지 아닌지 항상 체크를 하는 것을 확인했습니다. 저는 val 로 선언해서 사용했습니다.
멀티스레드의 상황에서 인스턴스에 스레드가 동시에 접근한다면 companion object를 쓰는게 맞겠지만 지금 상황에서는 그것을 논할 필요는 없을 것 같습니다.
companion object의 경우에는 클래스가 로드되는 시점에 메모리에 올라가게 됩니다. (static과 같은 시점이라고 하네요)

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/view/PlantListFragment.kt:
```diff
@@ -15,87 +15,128 @@ import com.example.samplekotlin.adpater.PlantListAdapter
-        for(i : Int in 0..16){
-            imgURL(plantImagesID[i], i, recyclerView)
+        for (i: Int in 0..16) {
+            imgURL(plantNames[i], i, recyclerView)
```
api 1개만 받고 결과로 받은 이미지로만 출력

YunSeokVV replied:
완료했습니다. 이번 커밋 내역에서 확인해보실수 있으실 겁니다.
늘 감사합니다.

YunSeokVV commented:
> 제코드 지워지면 그때 다음 depth로 나가시죠.

test

YunSeokVV committed:
- plantListAdapter 에서 리스너를 val 로 선언하고 어댑터 안에서 익명객체로 선언하는 방법이 떠올랐다. 좋은 방법인지 한번 여쭤보자.
- 싱글톤 패턴을 사용해서 Util 클래스에서 토스트 메세지를 동작시키는 것을 구현해봤다.
- 4주차 과제를 진행할때 executor관련 기능을 인터페이스에서 추상화 시키려고 했는데 추상 메소드의 본질을 이해하지 못하고 사용하였고, 디폴트 메소드로 만들어서 사용해버렸다.
(덤으로 동작도 안됐다)
- 사용하지 않는 보일러플레이트(19개의 식물 이름이 담겨있던 리스트)를 제거했다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/view/PlantListFragment.kt:
```diff
@@ -100,31 +78,19 @@ class PlantListFragment : Fragment() {
+                    data.add(plantVO)
                 }
 
+                plantlistAdapter.setClickListener(listener)
```
recyclerView 의 clickListener 와 adapter를 imgURL함수로 때낸 이유는 뭔가요? 초기화를 자주 해줘야하나요?

f-lab-loco replied:
notifyitemchanged() 를 공부해보세요. 파생되는 여러가지 method 들이 있는데 함께 공부 부탁드립니다.

YunSeokVV replied:
메소드로 따로 떼냈던 이유는 이전에 하드코딩된 문자열17개로 반복문을 돌려가면서 통신기능을 동작시켰었기 때문입니다. 
그리고 리사이클러뷰의 어댑터를 설정하는 코드는 반복문의 마지막 회차에서 설정시켰었는데 그냥 조건문 밖으로 빼놓기만 하면서 현재의 코드가 완성된 것입니다.

리사이클러뷰의 아이템이 변경됐다는 것을 알리는 다양한 메소드에 관해서 공부를 완료했습니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/view/PlantListFragment.kt:
```diff
@@ -100,31 +78,19 @@ class PlantListFragment : Fragment() {
                 }
 
+                plantlistAdapter.setClickListener(listener)
+                val recyclerView = recyclerVeiw
```
여기서 변수를 지정해서 썻는데 변수를 지정한 것과 line:59의 
`val recyclerView: RecyclerView = view.findViewById<RecyclerView>(R.id.plantListrecyclerView)`
에서 recyclerView로 쓰는 것과의 차이는 뭔가요?

f-lab-loco replied:
깊은 복사냐? 얕은 복사냐?
call by value & call by reference

YunSeokVV replied:
위에서 언급한것과 같은 내용입니다. 통신을 17번 반복해서 동작시키기 위해서 따로 메소드로 빼내는 작업을 했습니다.
두 recyclerView의 차이는 지역변수이냐 멤버변수이냐의 차이인데 call by reference 원리가 작동해서 지역변수로 선언했던 recyclerView 안에 멤버변수의 recyclerView가 담겼습니다. (두 객체가 서로 같은 주소를 바라보고 있기 때문에 같은 객체인 것으로 인식합니다.)

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/view/PlantListFragment.kt:
```diff
@@ -100,31 +78,19 @@ class PlantListFragment : Fragment() {
-                    val recyclerView = recyclerVeiw
-                    recyclerView.adapter = plantlistAdapter
-
+                for (i: Int in 0..9) {
```
forEach 를 쓰세요
response.body()?.results?.forEach() 이렇게 표기를 해보세요.

`val result = response.body()?.results ?: return`

YunSeokVV replied:
확인했습니다. 수정하도록 하겠습니다.

YunSeokVV committed:
5주차 수업때 받은 피드백들 수정
나의식물목록에서 좋아요 누른 식물 확인 가능

YunSeokVV committed:
MyGardenFragment에서  DB의 데이터를 읽어오는 liveData가 트랜잭션이 일어나는경우 옵저빙할 수 있게끔 구현 완료

YunSeokVV committed:
insert data with room + useCase, get data with retrofit + useCase

YunSeokVV committed:
- 상세화면에서 +버튼이 보여지는 로직을 수정
- 리사이클러뷰 아이템 롱클릭시 색상 변경 및 손을 뗐을때 원래 색상으로 돌아오게끔 구현
- 전체 식물목록 화면에서 필터링을 사용해 4개의 식물이나 전체 식물을 보여줄 수 있게끔 수정
- DB에 데이터 추가시 중복되는 데이터가 있더라도 앱이 죽지않게끔 수정
- DB에 데이터를 추가했을때 liveData가 옵저빙 하지 못했던 버그 수정
- DB에 데이터 추가시 primaryKey 가 중복되서 데이터가 제대로 저장되지 않았던 버그 수정
- 뷰의 visible속성을 Util메소드에서 설정하게끔 수정

YunSeokVV committed:
- 리사이클러뷰 클릭, 롱클릭 이벤트 코드 수정 (Gesture 삭제)
- 식물전체목록에서 필터링 기능을 구현할때 쓰는 라이브데이터 의 리스트를 얕은복사에서 깊은복사로 수정
- RoomDB 엔티티 클래스의 멤버변수 var -> val 로 수정
- 문자열 비교시 equals() 연산자에서 == 연산자로 수정.

YunSeokVV committed:
- 무의미한 주석/파일 전부 제거
- 리사이클러뷰 클릭이벤트 전부 어댑터의 생성자를 통해서 구현하게끔 설정
- PlantListFramgnet에서 비즈니스 로직을 ViewModel로 옮김
- 모든 viewModel을 get메소드가 아닌 public 접근자로 접근할 수 있게끔 변경

YunSeokVV committed:
- PlantListFramgnet의 라이브데이터가 옵저빙 하는 부분에서 어댑터 설정을 안하게 수정
- 뷰모델에서 어댑터 관련 코드가 동작됐는데 이를 다시 뷰에서 처리하게끔 수정.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/adpater/PlantListAdapter.kt:
```diff
@@ -1,39 +1,122 @@
-    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
-        var plantName: TextView = itemView.findViewById(R.id.plantName);
-        var plantImage: ImageView = itemView.findViewById<ImageView>(R.id.palntImage)
+    fun filterPlants() {
```
어뎁터 안에 이 function이 들어가는게 맞다고 생각하시나요?

YunSeokVV replied:
이번 수정에서 없앴습니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/adpater/PlantListAdapter.kt:
```diff
@@ -1,39 +1,122 @@
+            plantImage.clipToOutline = true
+
+            itemView.setBackgroundColor(ContextCompat.getColor(itemView.context, R.color.white))
+            if (longClickedItem.get(adapterPosition) == true) {
```
position을 들고 있는데 adapter안에 값이 추가되거나 삭제되면 꼬이는거아니에요?
그리고 recyclerView 테스트하실땐 적어도 20~30개 데이터를 넣어보고 재사용처리에 잘 하셨는지 체크 해야합니다.
여기 코드는 윗줄에 
```
itemView.setBackgroundColor(ContextCompat.getColor(itemView.context, R.color.white))
```
이게 있어서 괜찮지만 기본적으로 바로 위에 없다면 보기 어려울 것 같네요. pair로 else에 해당코드가 있어야 보기 쉽지 않을까요?

YunSeokVV replied:
슬랙에서 주고받은 대화인데, 위와같은 방식으로 개발을 하게되면 background 가 default 값이 설정되고 로직으로인해 한번 더 바뀌는건데 이는 분명 잘못된 부분입니다. (안바껴도 되는 색상이 두번 바뀌는 거니까요.)

이번 수정에서 이 문제도 해결했습니다. 감사합니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/adpater/PlantListAdapter.kt:
```diff
@@ -1,39 +1,122 @@
+
+    private var isFiltered = false
+
+    private val originalData = mutableListOf<Plant>()
```
original은 뭐고 _plantLiveData는 뭘까요? adapter에 리스트는 하나로 쓰는거아니에요? adpater에 비지니스로직이 들어가는 구조 같은데 맞게 설계하고 개발하신건가요?

YunSeokVV replied:
original 은 뷰모델로 옮겼습니다.

YunSeokVV committed:
- 롱클릭시 뷰홀더 타원형 깨짐 방지
- 리사이클러뷰 의도치않게 뷰가 재활용 되는 버그를 방지
- 어댑터에서 리스트를 하나만 사용하게끔 수정
- 필터링 관련 코드 비즈니스 로직을 뷰에서 사용하지 않게끔 수정
- 비즈니스 로직을 전부 뷰모델에서 사용하게끔 수정

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/viewmodel/PlantListFragmentViewModel.kt:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    fun filterPlantData(originalData: List<Plant>) : List<Plant> {
+        val filtered: List<Plant> = originalData.take(4).mapNotNull { it }
```
`maNotNull { it }` 은 null 인게 없는데 불필요 해보여요. 만약 이름이 null인걸 예외처리하고 싶다. 이러면 `maNotNull { it.name != null }` 이렇게 코딩 하면 좋을 것 같네요.

YunSeokVV replied:
네 숙지했습니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/viewmodel/PlantListFragmentViewModel.kt:
```diff
@@ -0,0 +1,67 @@
+        }
+
+    private val _isFiltered = MutableLiveData<Boolean>(false)
+    val isFiltered : LiveData<Boolean>get() {return _isFiltered}
```
code formatter, 정렬에 신경써주세요.

YunSeokVV replied:
정렬을 제가 못했네요... 주기적으로 하겠습니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/viewmodel/MainActivityViewModel.kt:
```diff
@@ -0,0 +1,26 @@
+
+    fun setfilterVisible() {
+        val currentvalue = _filterVisibleLiveData.value ?: true
+        _filterVisibleLiveData.value = !currentvalue
```
이 기능은 뭘까요? 단순히 observing 을 한번 더하는 코드일까요?

YunSeokVV replied:
우측상단의 필터링 버튼이 보이게할지 말지 여부를 나타내는 기능입니다.
모든 식물을 표현하는 Fragment에서 버튼이 보여져야 합니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/viewmodel/PlantListFragmentViewModel.kt:
```diff
@@ -0,0 +1,67 @@
+            }
+        }
+
+        return loadedList
```
밑에보면 이런 코드가 있더라구요?
```
plantListViewModel.setOriginalData(plantListViewModel.showRemotePlantData(data))
```
리턴으로 하기전에 originalData 를 넣어 줄 수 있지 않을까요?

YunSeokVV replied:
멘토님이 말씀해주신게 조금 더 나은 로직이군요. 참.. 이런걸 찾아내는게 저는 어렵던데 어떻게하면 좀 더 잘 찾을 수 있을지 잘 모르겠네요.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/viewmodel/PlantListFragmentViewModel.kt:
```diff
@@ -0,0 +1,67 @@
+    private val originalData = mutableListOf<Plant>()
+
+    fun setOriginalData(data: List<Plant>) {
+        this.originalData.clear()
```
_plants 이것과 originalData 를 분기해서 관리해야하나요?

YunSeokVV replied:
지난 수업시간에 이 부분에 관해서 따로 말씀해주셨던 내용이 기억이 나네요.
제가짰던 코드를 살펴보면서 고민을 해봤는데 제 생각의 결론은 분기해서 관리를 해야한다는 것입니다.
왜냐하면 _plants 라이브데이터가 서버에서 데이터를 받아오면 Plant 형식으로 받아오지 않기 때문입니다.

_plants 라이브 데이터는 서버에서 식물의 이미지 URL을 갖고 오면 해당 값을 옵저빙 하는게 목표입니다. 따라서 자료형은 아래와 같습니다. 
private val _plants = MutableLiveData<Response<UnsplashResults>>()

애초에 서버에서 받아왔을때 객체가 Plant 형식이면 멘토님 말씀이 맞다고 생각합니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/viewmodel/PlantListFragmentViewModel.kt:
```diff
@@ -0,0 +1,67 @@
+        val loadedList = mutableListOf<Plant>()
+        val resultList = response.body()?.results
+
+        resultList?.map {
```
map을 쓰실려고 노력했는데 부족해보이는것 같습니다.

바른 답은
```
return resultList?.mapNotNull {
            it.user.get("location")?.let{location->
               Plant(location.toString(), 0, it.urls["raw"].toString())
            }
        } ?: emptyList()
```
이렇게 하면 loadedList 를 안써도 되겠죠? 뭐가 다른지 고민해보세요.

또한 
it.urls["raw"] 이렇게 접근하면 "raw" 가 없을때 NullPointException 발생되지 않나요?
getOrNull을 주로 사용해주세요.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/view/PlantListFragment.kt:
```diff
@@ -1,69 +1,101 @@
+            // 원본으로 되돌린 경우
+            } else {
+                if (plantListViewModel.getOriginalData() != null) {
+                    plantlistAdapter.setData(plantListViewModel.getOriginalData())
```
빈리스트를 넣으면 문제가 되나요?? 비어져버린 List가 나와서 이렇게 예외처리한걸까요?

YunSeokVV replied:
위의 조건문은 예전에 코드를 수정하다가 만들었던 코드입니다. 앱을 처음 실행하면 서버에서 데이터를 받아오는 역할을 해주는 liveData를 else 문 안에서 실행하려고 하다가 NPE에러가 발생해서 추가했던 조건문입니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/view/PlantDetailFragment.kt:
```diff
@@ -0,0 +1,127 @@
+        wateringNeedsTextView.text = "every ${plant?.waterPeriod.toString()} days"
+
+        mainActivityViewModel.localPlant.observe(requireActivity(), Observer { data ->
+            data.forEach {
```
이 코드는 뭘까요? 설명좀 해주실수 있을까요? 불필요한 data forEach 문 같은데..

YunSeokVV replied:
네 forEach문은 위의 코드에서 굳이 필요가 없습니다. 왜냐하면 제가 작성한 CompanionUtil.likePlant메소드를 살펴보면 find 함수를 사용해서 list에서 특정값을 찾는 기능을 구현해놨기 때문에 굳이 forEach를 돌면서 원하는 정보를 찾을 필요가 없습니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/util/CompanionUtil.kt:
```diff
@@ -0,0 +1,22 @@
+class CompanionUtil {
+
+    companion object {
+        fun likedPlant(imgURL : String, localPlant : LiveData<List<Plant>>) : Int{
```
likedPlant인데 반환은 visible 이에요. 뭔가 이상하지않나요? 관련 Util이 얼마나 유용하게 쓰일지 잘 모르겠어요. 다방면에 응용될만한Util일까요?

YunSeokVV replied:
이 메소드는 뷰모델로 옮겼습니다. 말씀해주신대로 메소드명이나 안의 내용을 봤을때는 Util에 있기에 적합한 메소드는 아닌 것 같습니다.

f-lab-loco commented on app/src/main/java/com/example/samplekotlin/util/CompanionUtil.kt:
```diff
@@ -0,0 +1,22 @@
+            return if (localPlant.value?.find { it. imageResource == imgURL } != null) View.GONE else View.VISIBLE
+        }
+
+        fun makeToastMessage(context: Context, msg: String) {
```
나중에 `fun Context.makeToastMessage(message: String)`만들어보세요. 이걸 extension 이라고 합니다.
